Add some unit testing to internal/app/starter

gvallee opened the pull request:
**Description of the Pull Request (PR):**

Add some unit testing to `internal/app/starter/master_linux.go` and reorganize slightly the code to do so. Since I do not know whether modifying the code of `master_linux.go` to improve testing in the way I did it is acceptable, this is only a limited number of modifications and tests. If accepted, I will add more test cases in a separate future PR.

My modifications of `internal/app/starter/master_linux.go` intends to have shorter functions so testing is easier.

**This fixes or addresses the following GitHub issues:**

- Addresses #3315 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)


Attn: @singularity-maintainers

gvallee committed:
Add some unit testing to internal/app/starter/master_linux.go and reorganize slightly the code to do so

gvallee committed:
reorganize a test structure to please the linter and add an additional unit test and a slight modification of the original code to avoid a failure with the new unit tests

mem commented on internal/app/starter/master_linux.go:
```diff
@@ -20,93 +20,119 @@ import (
+	if isInstance && os.Getppid() == ppid {
+		// sleep a bit to see if child exit
+		time.Sleep(100 * time.Millisecond)
+
```
could you please delete this new newline? For this PR I would prefer to keep changes to a strict minimum.

mem commented on internal/app/starter/master_linux.go:
```diff
@@ -20,93 +20,119 @@ import (
+
+	go func() {
+		doRPCthread(rpcSocket, containerPid, engine, fatalChan)
+		runtime.Goexit()
```
what's the logic of keeping this here vs in the doRPCthread function? I can guess, but I would prefer to have it documented so that people looking at this in the future don't have to.

mem commented on internal/app/starter/master_linux.go:
```diff
@@ -20,93 +20,119 @@ import (
-func Master(rpcSocket, masterSocket int, isInstance bool, containerPid int, engine *engines.Engine) {
-	var fatal error
-	var status syscall.WaitStatus
+func doRPCthread(rpcSocket int, containerPid int, engine *engines.Engine, fatalChan chan error) {
```
I would prefer _not_ to call these functions "thread".

gvallee replied:
What do you want to call these functions then? I personally have no preference. Would just remove the term "thread" be okay?

mem replied:
I don't have a good name for them. I thought about suggesting something like "masterService" or "rpcService"; or "initRPCserver". But I'm not sure if those are good names either, and I didn't want to nudge anyone in a particular direction.

I know they are not threads, though.

cclerget replied:
@gvallee maybe just describe what they do ?

```
go createContainer(rpcSocket, containerPid, engine, fatalChan)
go startContainer(masterSocket, isInstance, containerPid, ppid, engine, fatalChan)
```

mem replied:
that's good from my point of view

mem commented on internal/app/starter/master_linux.go:
```diff
@@ -20,93 +20,119 @@ import (
 
-		runtime.Goexit()
-	}()
+func doMasterThread(masterSocket int, isInstance bool, containerPid int, ppid int, engine *engines.Engine, fatalChan chan error) {
```
same here

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+			name:         "fake engine struct",
+			rpcSocket:    -1,
+			containerPid: -1,
+			engine:       &fakeEngine,
```
please create a new engine (`&engines.Engine{}` or `new(engines.Engine)`) for each test case, otherwise the code under test might end up modifying the engine for the next test.

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+			shallPass:    false,
+		},
+		{
+			name:         "fake engine struct",
```
if I'm understanding correctly, this case is hitting the "invalid RPC socket" path. It's probably better to name the test something along those lines.

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+			shallPass:    false,
+		},
+		{
+			name:         "fake engine; fake rpcSocket",
```
not sure what this is testing. That the PID value of -1 is checked and "detected" as invalid?

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			go doRPCthread(tt.rpcSocket, tt.containerPid, tt.engine, fatalChan)
+			fatal = <-fatalChan
```
interesting... since the code does not have an obvious way to make it stop (IIRC there's a magic value that you can write to the RPC socket that will cause it to exit), it's not evident whether this test will hang here forever or not.

Maybe add a comment?

mem replied:
Maybe add a comment?

mem replied:
Let me try yet again: if I look at this test, it's not obvious that it will stop (i.e. that something is sent over `fatalChan`, partly because the name "fatalChan" suggest that would be the case only if a failure occurs).

Can you please add a comment clarifying that the test cases are such that there's always going to be a failure (both cases have `shallPass` set to `false`; someone reading this code might think there are other cases where `shallPass` is set to `true`) or that `createContainer` always sends something on `fatalChan` no matter what or whatever it is that guarantees that this test will not get stuck waiting for a value on `fatalChan`?

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+		t.Run(tt.name, func(t *testing.T) {
+			go doRPCthread(tt.rpcSocket, tt.containerPid, tt.engine, fatalChan)
+			fatal = <-fatalChan
+			if tt.shallPass && fatal != nil {
```
I wrote [this](https://github.com/sylabs/singularity/wiki/End-to-end-testing#check-all-cases) with E2E in mind, but it applies the same here. Can you please read that section and do what it says there regarding handling all cases?

mem replied:
Sorry, I don't see a response or a code change as suggested. Can you please follow the provided link?

mem replied:
I'm still waiting on an action regarding the original request.

mem replied:
good enough

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+			isInstance:   false,
+			containerPid: -1,
+			ppid:         -1,
+			engine:       &fakeEngine,
```
same

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+			isInstance:   false,
+			containerPid: -1,
+			ppid:         -1,
+			engine:       &fakeEngine,
```
same

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+			shallPass:    false,
+		},
+		{
+			name:         "fake engine",
```
same

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+			shallPass:    false,
+		},
+		{
+			name:         "fake engine; fake masterSocket",
```
same

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+		t.Run(tt.name, func(t *testing.T) {
+			go doMasterThread(tt.masterSocket, tt.isInstance, tt.containerPid, tt.ppid, tt.engine, fatalChan)
+			fatal = <-fatalChan
+			if tt.shallPass && fatal != nil {
```
same

mem replied:
and here the same comment: can you please follow the provided link?

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+			shallPass:    false,
+		},
+		{
+			name:         "fake engine",
```
same

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+			masterSocket: -1,
+			isInstance:   false,
+			pid:          -1,
+			engine:       &fakeEngine,
```
same

mem commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,171 @@
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			Master(tt.rpcSocket, tt.masterSocket, tt.isInstance, tt.pid, tt.engine)
```
probably better to omit this until there's something that can be tested? (and by "this" I mean "this test")

cclerget committed:
Move instance detachment in Singularity engine and get
rid of isInstance generic check in Master.
Add RPC/master socket checks while converting file descriptor
to *os.File pointer.
Remove now useless call to runtime.LockOSThread, it was
previously required to workaround issue with Go runtime <= 1.9.
Remove unit test for Master function for now and explaining why.
Remove engine nil checks, not reachable code path as there is a
check in main_linux.go for that.

cclerget reviewed (commented):
Changes committed directly in @gvallee branch after slack discussion. Waiting @mem feedbacks

gvallee committed:
Add comments about the channel that is used to synchronize threads.

sashayakovtseva commented on internal/app/starter/master_linux.go:
```diff
@@ -11,129 +11,131 @@ import (
+	}
+}
+
+// Master initializes a runtime engine and runs it
```
```suggestion
// Master initializes a runtime engine and runs it.
```

sashayakovtseva commented on internal/app/starter/master_linux.go:
```diff
@@ -11,129 +11,131 @@ import (
+	}
 
-	ppid := os.Getppid()
+	err = e.CreateContainer(containerPid, rpcConn)
```
Why there is no `runtime.LockOSThread` any more? 
In comments above I see you had some discussion with @cclerget in slack, was this discussed as well?

gvallee replied:
@sashayakovtseva This is incorrect, I did not discuss this point. The initial comment was that a reviewer (I do not remember who and it does not matter) did not like the naming I selected. @cclerget made a pass over the code to cleanup/ease the implementation of unit tests.
But are you asking here? The addition of comments?

sashayakovtseva commented on internal/app/starter/master_linux.go:
```diff
@@ -11,129 +11,131 @@ import (
-			fatalChan <- fmt.Errorf("failed to copy unix socket descriptor: %s", err)
-			return
-		}
+	rpcConn.Close()
```
Should this call be in defer? Just like in `startContainer`

jscook2345 replied:
Agreed. Should this be a deferred action?

cclerget replied:
The connection to RPC is closed only when container creation succeed in order to not execute stage 2 process (container process) in case of failure, then the master process will automatically killed the RPC process at exit

sashayakovtseva commented on internal/app/starter/master_linux.go:
```diff
@@ -11,129 +11,131 @@ import (
-		comm := os.NewFile(uintptr(masterSocket), "master-socket")
-		conn, err := net.FileConn(comm)
-		comm.Close()
+	// special path for engines which needs to stop before executing
```
```suggestion
	// special path for engines which need to stop before executing
```

gvallee replied:
That is actually incorrect, the subject is `special path` so it should be `needs` not `need` or it should be `special paths`

sashayakovtseva replied:
ah, that changes everything... I thought subject is ‘engines’  which should be stopped before start in order to execute prestart

sashayakovtseva commented on internal/app/starter/master_linux.go:
```diff
@@ -11,129 +11,131 @@ import (
+	// special path for engines which needs to stop before executing
+	// container process
+	if obj, ok := e.Operations.(interface {
+		PreStartProcess(int, net.Conn, chan error) error
```
Since this is a cleanup, maybe worth putting it into a separate type? e.g. function scoped interface

gvallee replied:
This is outside of the scope of this PR, which focuses on unit tests. I am not saying that we should not open an issue.

sashayakovtseva commented on internal/app/starter/master_linux.go:
```diff
@@ -11,129 +11,131 @@ import (
-			}
-			if err := obj.PreStartProcess(containerPid, conn, fatalChan); err != nil {
-				fatalChan <- fmt.Errorf("pre start process failed: %s", err)
+
```
```suggestion
```

sashayakovtseva commented on internal/app/starter/master_linux.go:
```diff
@@ -11,129 +11,131 @@ import (
+			// status will be set accordingly via MonitorContainer method below
+			sylog.Debugf("stage 2 process was interrupted, waiting status")
+			return
+		} else if data[0] == 'f' {
```
Can be simplified to `if data[0] == 'f'` since you have a return statement in previous if clause

gvallee replied:
The linter did not complain about this, I believe this is totally minor based on personal preferences.

sashayakovtseva commented on internal/app/starter/master_linux.go:
```diff
@@ -11,129 +11,131 @@ import (
+	err = e.PostStartProcess(containerPid)
+	if err != nil {
+		fatalChan <- fmt.Errorf("post start process failed: %s", err)
+		return
```
No need for this return, however for explicitness it is fine

sashayakovtseva replied:
Also a question, why `if isInstance && os.Getppid() == ppid` check is removed?

gvallee replied:
@sashayakovtseva  What is the request here? Add comments?

sashayakovtseva commented on internal/app/starter/master_linux.go:
```diff
@@ -11,129 +11,131 @@ import (
+	fatal := <-fatalChan
 
-	runtime.LockOSThread()
 	if err := e.CleanupContainer(fatal, status); err != nil {
```
Same question, why removing `runtime.LockOSThread`?

gvallee replied:
@sashayakovtseva I do not understand your request here, are you asking to add comments?

sashayakovtseva commented on internal/pkg/runtime/engines/singularity/process_linux.go:
```diff
@@ -425,7 +426,17 @@ func (engine *EngineOperations) PostStartProcess(pid int) error {
+		// send SIGUSR1 to the parent process in order to tell it
+		// to detach container process and run as instance.
+		// Sleep a bit in case child would exit
+		time.Sleep(100 * time.Millisecond)
```
Okay, looks like you and Cedric decided to move this here and thus `if isInstance && os.Getppid() == ppid` is removed in `master_linux.go`

gvallee replied:
@sasha I have nothing to do with it, I only wrote the unit test.

gvallee replied:
And I do not understand what this comment is about. Is it a question? A suggestion?

sashayakovtseva commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,118 @@
+				t.Fatalf("test %s expected to succeed but failed: %s", tt.name, fatal)
+			} else if !tt.shallPass && fatal == nil {
+				t.Fatalf("test %s expected to fail but succeeded", tt.name)
+			} else if tt.shallPass && fatal == nil {
```
Why having this elses? 
IMO [testify](https://github.com/stretchr/testify) is a perfect fit here.

gvallee replied:
@sasha Because it was required by others on the team regarding ALL unit tests in the ENTIRE codebase. It is supposed to improve readability and this topic is totally out-of-scope of this PR.

sashayakovtseva commented on internal/app/starter/master_linux_test.go:
```diff
@@ -0,0 +1,118 @@
+				t.Fatalf("test %s expected to succeed but failed: %s", tt.name, fatal)
+			} else if !tt.shallPass && fatal == nil {
+				t.Fatalf("test %s expected to fail but succeeded", tt.name)
+			} else if tt.shallPass && fatal == nil {
```
Same as above, testify will same time and make code cleaner IMO

gvallee replied:
@sasha this type of code organization has been requested by others; you ask for the opposite. Please sync with the other members of the team, decide what to do, but please do *not* PRs for these reasons.

gvallee committed:
Update internal/app/starter/master_linux.go

Co-Authored-By: Sasha Yakovtseva <sashayakovtseva@gmail.com>

gvallee committed:
Update internal/app/starter/master_linux.go

Co-Authored-By: Sasha Yakovtseva <sashayakovtseva@gmail.com>

gvallee committed:
Merge branch 'master' into issue_3315_masterlinux
